File: Leetcode/Top Interview Questions/Top_Interview_Questions_Easy/Array/IntersectionOfTwoArrays2.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> s;
        for(int i=0; i<nums1.size(); i++)
            s[nums1[i]]++;
        vector<int> result;
        for(int i=0; i<nums2.size(); i++)
            if(s.find(nums2[i]) != s.end()){
                s[nums2[i]]--;
                if(s[nums2[i]] == 0) s.erase(nums2[i]);
                result.push_back(nums2[i]);
            }
        return result;           
    }
};

// Time Complexity : O(m+n)
// Space Complexity: O(m)

File: LeetCode/Top Interview Questions/Top_Interview_Questions_Easy/Array/intersectionOfTwoArrays2.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) < len(nums2):
            nums1, nums2 = nums2, nums1
        seen = {}
        for i in nums1:
            if i not in seen:
                seen[i] = 1
            else:
                seen[i] += 1
        nums1 = []
        for j in nums2:
            if j in seen and seen[j] > 0:
                nums1.append(j)
                seen[j] -= 1
          
        return nums1
        
# Time Complexity - O(m + n)
# Space Complexity - O(max(m, n))
